test/unit/spec/narrative_core/upload/fileUploadWidget-413-custom-error-bad-spec.js

```diff
@@ -0,0 +1,90 @@
+define([
+    'jquery',
+    'kbase/js/widgets/narrative_core/upload/fileUploadWidget',
+    'base/js/namespace',
+    'narrativeConfig',
+    'testUtil',
+], ($, FileUploadWidget, Jupyter, Config, TestUtil) => {
+    'use strict';
+
+    /**
+     * Handles the case of the server emitting a 413 response which looks superficially
+     * like the custom response - "Request entity is too large" is in the response text -
+     * but is invalid JSON.
+     */
+
+    const fakeUser = 'notAUser',
+        filename = 'foo.txt',
+        stagingUrl = Config.url('staging_api_url') + '/upload';
+
+    describe('Test the fileUploadWidget 413 error with errors', () => {
+        beforeEach(function () {
+            jasmine.Ajax.install();
+            Jupyter.narrative = {
+                userId: fakeUser,
+                getAuthToken: () => {
+                    return 'fakeToken';
+                },
+                sidePanel: {
+                    $dataWidget: {
+                        $overlayPanel: {},
+                    },
+                },
+                showDataOverlay: () => {},
+            };
+            this.mockFile = new File(['0123456789'], filename, { type: 'text/plain' });
+
+            const responseText = 'oh dear, not JSON.';
+
+            jasmine.Ajax.stubRequest(stagingUrl).andReturn({
+                status: 413,
+                statusText: 'HTTP/1.1 413 OK',
+                contentType: 'application/json',
+                responseText,
+            });
+        });
+
+        afterEach(() => {
+            jasmine.Ajax.requests.reset();
+            jasmine.Ajax.uninstall();
+            TestUtil.clearRuntime();
+        });
+
+        it('Should display a default error message if the server reports the file is too big but the custom response is invalid JSON ', function (done) {
+            // Dropzone size unit is one mebibyte (1024 * 1024); uploadWidget takes bytes.
+            // For this test, size doesn't matter, as long as the max file size is greater
+            // than the fake file we send, as we don't want to trigger the dropzones
+            // max file size.
+            const maxFileSize = Math.pow(1024, 2);
+            const $wrapper = $('<div>');
+            const uploadWidget = new FileUploadWidget($wrapper, {
+                path: '/',
+                userInfo: {
+                    user: fakeUser,
+                    globusLinked: true,
+                },
+                maxFileSize,
+            });
+
+            uploadWidget.dropzone.addFile(this.mockFile);
+
+            // Realistically, this should poll until the condition is met, or
+            // timeout. Reason being that in real usage (or if using a real service mock),
+            // the call that generates the error may take an indeterminate amount of time,
+            // and certainly will not be on the next tick of the event loop.
```

Comment:
There's some tooling in `testUtil` that might be helpful to get the behavior you mention. The `waitForElementState` function might make the most sense here. Then you can make it an `async` test, `await` on that, and it'll fail if it times out.

Reply:
The existing file upload widget tests don't use that, I guess, since it was made later. If this works and is stable for now, that's fine.

Reply:
Yeah, it works, so I don't see a reason to just change it. 
It was more of a note to preserve the fact that it only works because it is using a specific style of mock service. It would fail, e.g., in an integration test, or a mock service which is either a true http server or includes a modest delay to simulate a network call.
I had just been through evaluating the implementation with different browsers. If memory serves, I found that some browsers wait for a good long while before reporting back the 413, others immediately, but in all cases it is a network call to the service occurs first, which is certainly longer in duration than a 0ms timeout.

Reply:
Okay, that sounds reasonable.